lib/raft/raft.hpp

```diff
@@ -0,0 +1,512 @@
+const auto ELECTION_TIMEOUT_MIN = 100;
+
+using NodeId = int;
+using Term = int;
+using Index = int;
+
+enum class NodeState {
+    Follower=0,
+    Candidate,
+    Leader
+};
+
+template <typename Operation>
+struct LogEntry {
+    Operation operation;
+    Term term;
+    Index index;
+};
+
+template <typename Operation, int N>
+class Log
+{
+    int m_size;
+    LogEntry<Operation> entries[N];
+
+    void remove_conflicting_entries(const LogEntry<Operation>* new_entries, int entry_count)
+    {
+        for (auto i = 0; i < size(); i++) {
+            for (auto j = 0; j < entry_count; j++) {
+                if (new_entries[j].index == entries[i].index &&
+                    entries[i].term < new_entries[j].term) {
+                    keep_until(i);
+                    return;
+                }
+            }
+        }
+    }
+
+public:
+    Log() : m_size(0)
+    {
+    }
+
+    int size() const
+    {
+        return m_size;
+    }
+
+    void append(LogEntry<Operation> entry)
+    {
+        if (m_size < N) {
+            entries[m_size] = entry;
+            m_size ++;
+        } else {
+            WARNING("log is already full");
+        }
+    }
+
+    LogEntry<Operation>& operator[](int i)
+    {
+        return entries[i];
+    }
+
+    Index last_index() const
+    {
+        if (m_size > 0) {
+            return entries[m_size - 1].index;
+        }
+
+        return 0;
+    }
+
+    Term last_term() const
+    {
+        if (m_size > 0) {
+            return entries[m_size - 1].term;
+        }
+
+        return 0;
+
+    }
+
+    LogEntry<Operation>* find_entry(Term term, Index index)
+    {
+        for (auto i = 0; i < size(); i++) {
+            if (entries[i].index == index && entries[i].term == term) {
+                return &entries[i];
+            }
+        }
+
+        return nullptr;
+    }
+
+    void merge(const LogEntry<Operation>* entries, int entry_count)
+    {
+        remove_conflicting_entries(entries, entry_count);
+
+        for (auto i = 0; i < entry_count; i++) {
+            if (entries[i].index > last_index()) {
+                append(entries[i]);
+            }
+        }
+    }
+
+    // Discards all elements after the nth
+    void keep_until(int n)
+    {
+        m_size = n;
+    }
+};
+
+template <typename StateMachine>
+struct Message {
+    enum class Type {
+        VoteRequest=0,
+        VoteReply,
+        AppendEntriesRequest,
+        AppendEntriesReply,
+    };
+
+    Type type;
+    Term term;
+    NodeId from_id;
+
+    union {
+        struct {
+            Index last_log_index;
+            Term last_log_term;
+        } vote_request;
+
+        struct {
+            bool vote_granted;
+        } vote_reply;
+
+        struct {
+            int count;
+            Index leader_commit;
+            Term previous_entry_term;
+            Index previous_entry_index;
+            // TODO: How many entries
+            LogEntry<typename StateMachine::Operation> entries[10];
```

Comment:
Consider wrapping `10` in a const variable at the top of the file together with the TODO note